Fix documentation of default from missing with marshmallow 3

lafrech opened the pull request:
Addresses #342.

It sucks not being able to document default values automatically, especially since it would probably work most of the times, but I don't see a safe way to do it.

Without this feature removal, we may pass to the spec dict objects that are not json serializable, which will trigger an exception when serializing the spec file.

lafrech committed:
Fix documentation of default from missing with marshmallow 3

sloria commented:
> OTOH, in most cases, dumping missing is perfectly fine, so removing automatic default documentation would be a regression.

This. As a user, I want apispec to work mostly transparently, i.e. without having to modify my app code. 

Maybe we could handle errors better to lead the user toward using `doc_default` when it's necessary, but I don't think we should degrade the 90% use case.

lafrech commented:
Well, we could try to dump the `missing` value with the `Field`. Let's give it a try.

I don't know if we should document the `missing` value of `Nested` fields or container fields with `Nested` inside. The examples I found for default values are "scalar" values like integers or strings but I'm not sure a dict is allowed. Anyway, that's nothing new. Same applies to MA2.

lafrech commented:
I just pushed a new commit to serialize `missing` with the field.

It has a few smells:

- Accesses underscored `_serialize` method (no big deal).
- Can't pass `attr` and `obj` parameters of `_serialize`. This shouldn't be an issue in most cases as those are almost never used. In fact, the only uses I know of are `Function` and `Method` (I already commented about this in https://github.com/marshmallow-code/marshmallow/issues/1046) and those are dump_only.

We should probably only document missing for fields that are not dump_only. It doesn't make sense anyway, does it?

Also, I'd need to add proper tests to show it works with a field where this matters, such as `DateTime`.

lafrech commented:
Not quite.

Indeed we shouldn't document default on a `dump_only` field.

However, `Function` and `Method` may not be `dump_only` and in this case, the code is likely to crash. But maybe not, if the method can take `None` as a parameter.

> Maybe we could handle errors better to lead the user toward using `doc_default` when it's necessary

So shall we call `_serialize`, catch `Exception` and return a message telling the user to use `doc_default`? The result may even be wrong without an exception being raised, so this wouldn't be 100% sure. But admittedly, those would be corner cases. Perhaps a note in the docs would be sufficient.

Or rather than passing `None`, could we pass some sort of proxy object as `attr` and `obj` that would raise an exception when used? For `obj`, we could override getter methods, etc. But for `attr`...

sloria commented:
For now, I think documentation may be the best course.

lafrech committed:
Update CHANGELOG

lafrech committed:
Update docs about MarshmallowPlugin properties documentation

lafrech commented:
I added a piece of doc about `MarshmallowPlugin`.

Feedback welcome.

> Indeed we shouldn't document default on a `dump_only` field.

Maybe not but a field with `dump_only` and `missing` is wrong in marshmallow anyway, and we generally don't try to validate user inputs in apispec, so I removed the corresponding case from this PR.

Note: `default` in `_VALID_PROPERTIES` is useless because it can't be in `metadata` as it is a reserved marshmallow attribute. I thought of removing it from the list but things might change in https://github.com/marshmallow-code/marshmallow/issues/1350 (MA 4) so I left it there for now, it is harmless.

sloria committed:
Revise docs

sloria commented:
Go ahead and merge when you think this is ready.

lafrech commented:
@sloria, @bangertm, anything else you'd like to add to 3.0 ?

sloria commented:
nothing from me

Bangertm commented:
Been checked out the past few days, but I like the recent changes. The only other thing I have been thinking about is extracting the logic for finding Schemas in OpenAPI objects from `MarshmallowPlugin` and/or improving the documentation of those methods. 

`resolve_parameters`, `resolve_schema_in_request_body`, `resolve_schema` and `OpenAPIConverter.resolve_schema_dict` all have to do with searching for schemas within various OpenAPI objects and then converting them via `OpenAPIConverter`. It might be convenient if all this logic was implemented in a mixin class so that `MarshmallowPlugin` was left with just the plugin methods and a couple of helpers. 

It would also be convenient if the documentation for these methods was a little more clear about the expected structure of the input and where schemas could go. Part of the rational for pulling all the logic out would be to improve documentation and understandably, but we could also just improve the documentation of the methods as they are now.

sloria commented:
> resolve_parameters, resolve_schema_in_request_body, resolve_schema and OpenAPIConverter.resolve_schema_dict all have to do with searching for schemas within various OpenAPI objects and then converting them via OpenAPIConverter. It might be convenient if all this logic was implemented in a mixin class so that MarshmallowPlugin was left with just the plugin methods and a couple of helpers.

To follow the pattern in https://github.com/marshmallow-code/apispec/pull/493, we could do

```python
class MarshmallowPlugin(BasePlugin):
    SchemaResolver = SchemaResolver

class MyMarshmallowPlugin(MarshmallowPlugin):
    class SchemaResolver(MarshmallowPlugin.SchemaResolver):
        def resolve_parameters(...):
            ...
```
